include/flox/core/exceptions.hh

```diff
@@ -74,76 +74,150 @@ enum error_category {
 /** Typed exception wrapper used for misc errors. */
 class FloxException : public std::exception
 {
-private:
 
-  /** Corresponds to an error_category, in this case `EC_FLOX_EXCEPTION`. */
-  static constexpr std::string_view categoryMsg = "general error";
+private:
 
   /** Additional context added when the error is thrown. */
   std::optional<std::string> contextMsg;
 
   /**
-   * If some other exception was caught before throwing this one, caughtMsg
+   * If some other exception was caught before throwing this one, @a caughtMsg
    * contains what() of that exception.
    */
   std::optional<std::string> caughtMsg;
 
+  /** The final what() message. */
+  std::string whatMsg;
+
+
 public:
 
   explicit FloxException( std::string_view contextMsg )
     : contextMsg( contextMsg )
-  {}
+  {
+    this->whatMsg = std::string( this->getCategoryMessage() ) + ": "
+                    + std::string( contextMsg );
+  }
 
-  explicit FloxException( std::string_view contextMsg, const char *caughtMsg )
+  explicit FloxException( std::string_view contextMsg,
+                          std::string_view caughtMsg )
     : contextMsg( contextMsg ), caughtMsg( caughtMsg )
-  {}
+  {
+    this->whatMsg = std::string( this->getCategoryMessage() ) + ": "
+                    + std::string( contextMsg );
+  }
+
+  /* For child classes. */
+  explicit FloxException( std::string_view           categoryMsg,
+                          std::optional<std::string> contextMsg,
+                          std::optional<std::string> caughtMsg )
+    : contextMsg( contextMsg ), caughtMsg( caughtMsg ), whatMsg( categoryMsg )
+  {
+    if ( contextMsg.has_value() ) { this->whatMsg += ": " + ( *contextMsg ); }
+    if ( caughtMsg.has_value() )
+      {
+        assert( contextMsg.has_value() );
+        this->whatMsg += ": " + ( *caughtMsg );
+      }
+  }
+
 
   [[nodiscard]] virtual error_category
   getErrorCode() const noexcept
   {
     return EC_FLOX_EXCEPTION;
   }
 
+  [[nodiscard]] std::optional<std::string>
+  getContextMessage() const noexcept
+  {
+    return this->contextMsg;
+  }
+
+  [[nodiscard]] std::optional<std::string>
+  getCaughtMessage() const noexcept
+  {
+    return this->caughtMsg;
+  }
```

Comment:
In this class it doesn't matter; but it's C++ _best practice_ to make all member variables private for `classes`, and interact with them using `getFoo` and `setFoo` functions.
In cases where you really want an escape hatch we have `friend` declarations; but in general it's best to use those sparingly ( as code-bases grow people make certain assumptions about private/public fields' and whether there's any _special_ related setup in CTX or some similar update operration ).

For _plain_ records where we want public fields, we use `struct` instead of `class`.

`clang-tidy` will gripe at us for public fields in classes.